### ISSUE TITLE
PoolCluster stops queuing and fails with ETIMEDOUT

### ISSUE THREAD

**Issue by unlucio** [mysqljs/mysql#904]
Hello everybody,
I'm still investigating on this headache but I'd greatly appreciate whatever suggestion.

I'm using a PoolCluster with a 10 connection limit per pool, selecting a pool and running N queries on it.
I'm inserting a sample of 300k rows, at 1k rows per query (using the multiple queries feature).
I already checked the generated SQL and it is fine.
Everything works fine as long as I try to send 30 (exactly 30) subsequent queries (30k rows):
10 connections get pulled out of the pool, the other get queue and performed as soon as a connection is available. And up to this point everything is just great :)

Running a test with 31 connection (30500 rows): everything goes wreck :/
the whole cluster tumbles down as the 1st `getConnection()` receives a `ETIMEDOUT` error and all the subsequent `getConnection()` call receive a `Error: Pool is closed` error. [Here](http://pastebin.com/Ndy2DMiL) you can find a sample error log with stack traces.
I added a `console.log()` (and marked it as `mysql lib stack:  Error:`) in the library as suggested [by this answer](https://github.com/felixge/node-mysql/issues/879#issuecomment-49813171) (the case did look kind of similar to mine)

Increasing config variables values such as  `connectionLimit`, `acquireTimeout`, `removeNodeErrorCount` and explicitally setting `waitForConnections`, `canRetry` to true didn't help either.

Checking the server's configuration, the numbers seem fine:

```
mysql> SHOW VARIABLES LIKE "%connect%";
+-----------------------------------------------+--------------------------------------------+
| Variable_name                                 | Value                                      |
+-----------------------------------------------+--------------------------------------------+
| character_set_connection                      | utf8                                       |
| collation_connection                          | utf8_general_ci                            |
| connect_timeout                               | 10                                         |
| disconnect_on_expired_password                | ON                                         |
| init_connect                                  | SET collation_connection = utf8_general_ci |
| max_connect_errors                            | 100                                        |
| max_connections                               | 151                                        |
| max_user_connections                          | 0                                          |
| performance_schema_session_connect_attrs_size | 512                                        |
+-----------------------------------------------+--------------------------------------------+
9 rows in set (0.00 sec)
```

Shouldn't the pool/poolCluster feature manage these kind of situations by its self?
What would be the best way to manage this kind of error and reconnect the lost pool?

As said I'm still investigating this situation and I'll try to add more information as soon as they come out. I'm sorry I'm not able to attach a fiddle test case at this time.

**Comment by DasithKuruppu** [comment]
Im stuck on this same issue too , i tried on linux as well as windows, it seems to give the same error and it used to run just fine a while back it hangs up when connecting to my 2nd database, are u guys on a solution to this yet ?

**Comment by dougwilson** [comment]
@DasithKuruppu I haven't been able to reproduce and so no work has been done. If someone can suggest a fix who can reproduce and verify it, that would be great :)

**Comment by DasithKuruppu** [comment]
module.exports = function (mysql) {
var config=require('./Config.js');

var sourcepool = mysql.createPool(config.databases.source);
var destinationpool=mysql.createPool(config.databases.destination);

// Create 2 mysql pools to interact with 2 databases in 2 different locations
sourcepool.getConnection(function(err,sourceDB_Connection){
    console.log("connected to sourceDB");
 if(err) { console.log(err); return; }
    destinationpool.getConnection(function(err,destinationDB_Connection){
        console.log("connected to Destinationdb");
        if(err) { console.log(err); return; }
});
});

Here basically im using express (latest version) and node-mysql(latest version) it returns destination DB as undefined , the two pool connections are to two databases hosted online(and yea i can connect to both dbs using phpmyadmin) so both databases are up , funney thing is this same code ran fine for the last couple of weeks and today suddenly it hangs with this error.

http://prntscr.com/6whqom

hope this helps in figuring it out

**Comment by dougwilson** [comment]
@DasithKuruppu no, it doesn't help. I've already tried similar code. The issue is environmental, which is why I'm not able to reproduce since I'm not able to reproduce your environment.

**Comment by DasithKuruppu** [comment]
Yea which is why i tried the same code on my linux(elementary freya) with a fresh node installation, seemed to reproduce the same error ?

**Comment by dougwilson** [comment]
I assume you're still connecting to the same MySQL host, correct?

**Comment by DasithKuruppu** [comment]
yes

**Comment by DasithKuruppu** [comment]
I could give you access to my two Db's hosted online if you want , could you give me an email or something i can mail u the logins

**Comment by dougwilson** [comment]
Ok. That's still your own environmental. All you did was change the client host, but you still kept those variables the same. My email is listed on my GitHub profile.

**Comment by DasithKuruppu** [comment]
Just mailed u the details , hope that helps u figure it out

**Comment by dougwilson** [comment]
Thank you much :) I got the email and I'll look at it today and hopefully moves this ticket forward :)!

**Comment by DasithKuruppu** [comment]
Thanks a lot , and hope you keep me updated :+1:

**Comment by varmint09** [comment]
Was there a solution found to this, this is the same issue I have...

**Comment by dougwilson** [comment]
Hi @varmint09 , unfortunately not. I need help trying to reproduce the issue or if someone can debug the issue when they have it and either offer up a PR or a detailed description for someone else (like myself) to create a fix.

**Comment by varmint09** [comment]
Hi @dougwilson looks to be quite a common issue, looks like the last two issues posted are similar.
I have tried using mysql2, but that didnt work either, and im currently in a situation where a site cant go live as amazon RDS cant be connected to...

Not quite sure how to proceed or help really, as this is critical for me!

**Comment by sidorares** [comment]
hi @varmint09 

- do you see same problem if you use smaller queries ( say, 1 statement instead of multiple 1000 per single command ) ?
- if the goal is to dump lots of data from source, transform it and insert to destination than it might be more efficient to use `LOAD DATA LOCAL INFILE` using generated sql as input

also, it might be helpful to see logs with connection `debug` flag set to true

**Comment by varmint09** [comment]
These are very modest queries indeed, but no query gets to run as the database connection is not established correctly.

If I decided not to use connection pools would that make any difference? Anything else I can try?

**Comment by sidorares** [comment]
have you tried using one direct connection @varmint09  ?

**Comment by varmint09** [comment]
Just tried, no luck. @sidorares

**Comment by sidorares** [comment]
and standard cli client works fine from that machine?

**Comment by unlucio** [comment]
@varmint09 we stubbornly tried for a year to use node-mysql experiencing timeouts, protocol_sequence_errors, "left over connections" showing as pending on the server for hours and other fatal errors and high memory consumption.
During the last year we investigated many times a number of problems and odd behaviours  and general instability that seems to emerge when working with tenth of thousands of queries like the following (example)
```javascript
mysql.format('INSERT INTO ' + database + ' (' + fields + ') VALUES ? ON DUPLICATE KEY UPDATE ' + queryOptions.updateString, item);

// items of about 40 properties
```
or when running queries generating results of hundreds of thousands or million records of ~40, ~50 properties (using the stream interface).
Using both pools (10 connections) or a single persistent connection always experiencing problems. I really wish in this year we were able to come up with a consistent example to investigate the issues but unfortunately they were quite difficult to reproduce in a concise way if not running the whole system and pushing it to it's live workload on our dev machines as well.
We switched for good about a month ago to mysql2 and every thing now works fine and the memory consumption lowered of about 40% as well (and our code is the same, we really just switched the "driver").

Just a personal note: I would have rather sticked with node-mysql since it seems to be the most used one and I'm not a fan of dispersing people and energy of OS projects that kind of make the same thing. I wish I could make a contribution on this one and solve our problems at the same time but unfortunately I wasn't able to wrap my mind around the project's code and structure and and I don't have a in-depth enough knowledge of mysql to be able to "sense" what's going on. The only thing I can really do is report my experience and maybe try to see if there are any chance to show a live system experiencing those issues to someone that might be able to understand more.

**Comment by varmint09** [comment]
@unlucio My understanding is that you are also using ELB, and connecting to a MySQL RDS instance, I will try using MySQL2, and I just wanted to check if that worked for you?

My config object I am passing into createPool is 

  {
        "host" : process.env.RDS_HOSTNAME,
        "port" : +process.env.RDS_PORT,
        "user" : process.env.RDS_USERNAME,
        "password" : process.env.RDS_PASSWORD,
        "database" : process.env.RDS_DATABASE_NAME,
        "ssl" : "Amazon RDS"
  }

I will try again using mysql2 in the interim, and hopefully that will work for me too.

@dougwilson seems like quite a big bug to not mention somewhere more prominently, as from a glance through issues this looks quite common...

**Comment by dougwilson** [comment]
Hi @varmint09 , there still isn't anything we can do. My comment at https://github.com/felixge/node-mysql/issues/904#issuecomment-140587694 has still be unanswered. Can you help with that? We need detailed instructions on how to reproduce the issue or there is no hope of us being able to fix anything (how can we fix something we cannot see?). Another option is to debug your issue, determine the solution, and propose a fix in a PR.

**Comment by unlucio** [comment]
@varmint09 yes we are indeed using ELB and RDS and yes moving to mysql did improve a lot our experience.

**Comment by alanclarke** [comment]
+1

**Comment by onsamp** [comment]
+1

**Comment by clivend** [comment]
+1

**Comment by takion** [comment]
+1

**Comment by fakhruz3105** [comment]
+1
